### PR TITLE
transport: introduce `git_transport_smart_remote_connect_options`

### PR DESCRIPTION
**Pull request by lhchavez**
6fc6eeb66c40310086c8f059cae41de69ad4c6da removed
`git_transport_smart_proxy_option`, and there was nothing added to
replace it. That made it hard for custom transports / smart
subtransports to know what remote connect options to use (e.g. proxy
options).

This change introduces `git_transport_smart_remote_connect_options` to
replace it.

### CODE REVIEW

**Comment by nmeum** [comment]
Has there been any progress on this? My understanding is that git2go is not capable of supporting libgit2 1.4.X without these changes which is a bit of a bummer for distributions as it is a blocker for upgrading libgit2 if the distribution includes packages which use git2go (see https://github.com/macports/macports-ports/pull/14602 for example). As such, it would be cool to get this merged and maybe even make a new libgit2 1.4.4 release with these changes included.

File: src/libgit2/transports/smart.c [hunk 1]
```diff
@@ -425,6 +425,18 @@ int git_transport_smart_credentials(git_credential **out, git_transport *transpo
 	return connect_opts->callbacks.credentials(out, t->url, user, methods, connect_opts->callbacks.payload);
 }
 
+int git_transport_smart_remote_connect_options(
```
Comment (ethomson) [inline src/libgit2/transports/smart.c:428]:
```suggestion
int git_transport_remote_connect_options(
```

File: include/git2/sys/transport.h [hunk 2]
```diff
@@ -270,6 +271,19 @@ GIT_EXTERN(int) git_transport_smart_credentials(git_credential **out, git_transp
...
+/**
+ * Get a copy of the remote connect options
+ *
+ * All data is copied and must be freed by the caller by calling
+ * `git_remote_connect_options_dispose`.
+ *
+ * @param out options struct to fill
+ * @param transport the transport to extract the data from.
+ */
+GIT_EXTERN(int) git_transport_smart_remote_connect_options(
```
Comment (ethomson) [inline include/git2/sys/transport.h:283]:
```suggestion
GIT_EXTERN(int) git_transport_remote_connect_options(
```

File: tests/libgit2/transport/register.c [hunk 3]
```diff
@@ -77,3 +79,146 @@ void test_transport_register__custom_transport_ssh(void)
...
+		git_smart_service_t action)
+{
+	struct custom_subtransport *t = (struct custom_subtransport *)transport;
+	git_remote_connect_options opts = GIT_REMOTE_CONNECT_OPTIONS_INIT;
+	int ret;
+
+	GIT_UNUSED(url);
+	GIT_UNUSED(action);
+
+	ret = git_transport_smart_remote_connect_options(&opts, t->owner);
```
Comment (ethomson) [inline tests/libgit2/transport/register.c:137]:
```suggestion
	ret = git_transport_remote_connect_options(&opts, t->owner);
```

File: include/git2/sys/remote.h [hunk 4]
```diff
@@ -26,6 +28,17 @@ typedef enum {
 	GIT_REMOTE_CAPABILITY_REACHABLE_OID = (1 << 1),
 } git_remote_capability_t;
 
+/**
+ * Disposes libgit2-initialized fields from a git_remote_connect_options.
```
Comment (ethomson) [inline include/git2/sys/remote.h:32]:
```suggestion
 * Disposes libgit2-initialized fields from a git_remote_connect_options.
 * This should only be used for git_remote_connect_options returned by
 * git_transport_remote_connect_options.
```

**Review by ethomson — COMMENTED** [review]
Let's shorten the name - we're getting the `remote_connect_options` on the `git_transport`.

File: include/git2/sys/transport.h [hunk 5]
```diff
@@ -261,6 +261,29 @@ GIT_EXTERN(int) git_transport_smart_certificate_check(git_transport *transport,
...
+/**
+ * Get a copy of the proxy options
+ *
+ * The url is copied and must be freed by the caller.
+ *
+ * @param out options struct to fill
+ * @param transport the transport to extract the data from.
+ */
+GIT_EXTERN(int) git_transport_smart_proxy_options(git_proxy_options *out, git_transport *transport);
+
```
Comment (ethomson) [inline include/git2/sys/transport.h:273]:
```suggestion
```

I think that this snuck in and should be removed.

**Comment by ethomson** [comment]
Thanks @lhchavez !

**Comment by xiaket** [comment]
Thanks @lhchavez . Is it possible to have a release for this change @ethomson ? Downstream repos are really looking forward to it.

**Pull request merged**